### PR TITLE
Practice2

### PR DESCRIPTION
**Pull request by Osievskaya23**

### CODE REVIEW

File: immutableclass/ImmutableClassTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        cat.getFamily().setMother("mother Dog");
+        System.out.println("\nCat after changing family:" +
+                "cat name = " + cat.getName() + ";\n" +
+                "cat age = " + cat.getAge() + ";\n" +
+                "cat's father = " + cat.getFamily().getFather() + ";\n" +
+                "cat's mother = " + cat.getFamily().getMother() + ";");
+    }
+}
+
+final class ImmutableAnimal {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: missed class modifier

File: immutableclass/ImmutableClassTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    private final String name;
+    private final int age;
+    private final Family family;
+
+    ImmutableAnimal(String name, int age, Family family) {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: missed constructor midifier

File: immutableclass/ImmutableClassTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    String getName() {
+        return name;
+    }
+
+    int getAge() {
+        return age;
+    }
+
+    Family getFamily() {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: missed method modifier

File: immutableclass/ImmutableClassTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        //this.family = family;
+        Family familyClone = new Family(family.getFather(), family.getMother());
+        this.family = familyClone;
+    }
+
+    String getName() {
+        return name;
+    }
+
+    int getAge() {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: missed method modifier

File: immutableclass/ImmutableClassTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    ImmutableAnimal(String name, int age, Family family) {
+        this.name = name;
+        this.age = age;
+        //this.family = family;
+        Family familyClone = new Family(family.getFather(), family.getMother());
+        this.family = familyClone;
+    }
+
+    String getName() {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: missed method modifier

File: immutableclass/ImmutableClassTest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        return age;
+    }
+
+    Family getFamily() {
+        //return family;
+        return new Family(this.family.getFather(), this.family.getMother());
+    }
+}
+
+class Family {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: missed class modifier

File: immutableclass/ImmutableClassTest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        return name;
+    }
+
+    int getAge() {
+        return age;
+    }
+
+    Family getFamily() {
+        //return family;
+        return new Family(this.family.getFather(), this.family.getMother());
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: if you need to return mutable object will better if your class family implements interface cloneable and override method clone.

File: immutableclass/ImmutableClassTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    }
+}
+
+class Family {
+
+    private String father;
+    private String mother;
+
+    Family(String father, String mother) {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: modifier

File: immutableclass/ImmutableClassTest.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        this.father = father;
+        this.mother = mother;
+    }
+
+    String getFather() {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: modifier

File: immutableclass/ImmutableClassTest.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        return father;
+    }
+
+    void setFather(String father) {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: modifier

File: immutableclass/ImmutableClassTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        this.father = father;
+    }
+
+    String getMother() {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: modifier

File: immutableclass/ImmutableClassTest.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        return mother;
+    }
+
+    void setMother(String mother) {
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: modifier

File: immutableclass/ImmutableClassTest.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        this.mother = mother;
+    }
+}
```
Comment (BondarenkoJek) [inline immutableclass/ImmutableClassTest.java]: after last braces should be 1 empty line

File: serialization/Circle.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
+package serialization;
+
+public class Circle implements Figure {
+    public static final double PI = 3.141_592_653_589_793;
+    private String name;
+    private int radius;
+    public double square;
+
+    Circle(String name, int radius) {
```
Comment (BondarenkoJek) [inline serialization/Circle.java]: modifier

File: serialization/Figure.java [hunk 15]
```diff
@@ -0,0 +1,8 @@
+package serialization;
+
+public interface Figure {
+    double calculateSquare();
+    String getName();
+    double getSquare();
+}
+
```
Comment (BondarenkoJek) [inline serialization/Figure.java]: delete unnecessary empty line

File: serialization/Group.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class Group {
+    private String name;
+    private List<Figure> figures = new ArrayList<>();
+    private List<Group> groups = new ArrayList<>();
+
+    Group(String name) {
```
Comment (BondarenkoJek) [inline serialization/Group.java]: modifier

File: serialization/Group.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+    public List<Figure> getFigures() {
+        return figures;
+    }
+
+    public List<Group> getGroups() {
+        return groups;
+    }
+
+    void addFigures(Figure ... neadAddingFigures) {
```
Comment (BondarenkoJek) [inline serialization/Group.java]: modifier

File: serialization/Group.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        figures.addAll(Arrays.asList(neadAddingFigures));
+    }
+
+    void addGroups(Group ... neadAddGroups) {
```
Comment (BondarenkoJek) [inline serialization/Group.java]: modifier

File: serialization/Group.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        groups.addAll(Arrays.asList(neadAddGroups));
+    }
+
+
+    void getFigureSquares() {
```
Comment (BondarenkoJek) [inline serialization/Group.java]: modifier

File: serialization/Square.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
+package serialization;
+
+public class Square implements Figure {
+    public String name;
+    private int a;
+    public double square;
+
+    Square(String name, int a) {
```
Comment (BondarenkoJek) [inline serialization/Square.java]: modifier

File: serialization/Triangle.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
+package serialization;
+
+public class Triangle implements Figure {
+    public String name;
+    private int a;
+    private int b;
+    private int c;
+    public double square;
+
+    Triangle(String name, int a, int b, int c) {
```
Comment (BondarenkoJek) [inline serialization/Triangle.java]: modifier

File: serialization/encoder/JSONEncoder.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+    private StringBuilder getTabs(int tabs) {
+        StringBuilder tab = new StringBuilder();
+        while (tabs != 0) {
+            tab.append("\t");
+            tabs--;
+        }
+        return tab;
+    }
+}
+
```
Comment (BondarenkoJek) [inline serialization/encoder/JSONEncoder.java]: delete line

File: serialization/encoder/XMLEncoder.java [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+    private StringBuilder getMethodPropertyLine(Class clazz, Method m, int tabs) {
+        StringBuilder string = new StringBuilder();
+        string.append(getTabs(tabs));
+        string.append("<void method=\"");
+        string.append(m.getName());
+
+        string.append("\"> </void>\n");
+
+        //string.append("\">\n");
+        //string.append(getTabs(tabs));
```
Comment (BondarenkoJek) [inline serialization/encoder/XMLEncoder.java]: delete comment code

File: immutableclass/ImmutableClassTest.java (hunk 1 above)
Comment (Osievskaya23) [reply immutableclass/ImmutableClassTest.java]: refactored

File: immutableclass/ImmutableClassTest.java (hunk 2 above)
Comment (Osievskaya23) [reply immutableclass/ImmutableClassTest.java]: refactored

File: immutableclass/ImmutableClassTest.java (hunk 3 above)
Comment (Osievskaya23) [reply immutableclass/ImmutableClassTest.java]: refactored

File: immutableclass/ImmutableClassTest.java (hunk 4 above)
Comment (Osievskaya23) [reply immutableclass/ImmutableClassTest.java]: refactored

File: immutableclass/ImmutableClassTest.java (hunk 5 above)
Comment (Osievskaya23) [reply immutableclass/ImmutableClassTest.java]: refactored

File: serialization/Group.java (hunk 19 above)
Comment (Osievskaya23) [reply serialization/Group.java]: refactored

File: serialization/Square.java (hunk 20 above)
Comment (Osievskaya23) [reply serialization/Square.java]: refactored

File: serialization/Triangle.java (hunk 21 above)
Comment (Osievskaya23) [reply serialization/Triangle.java]: refactored

File: serialization/encoder/JSONEncoder.java (hunk 22 above)
Comment (Osievskaya23) [reply serialization/encoder/JSONEncoder.java]: refactored

File: serialization/encoder/XMLEncoder.java (hunk 23 above)
Comment (Osievskaya23) [reply serialization/encoder/XMLEncoder.java]: refactored

File: serialization/encoder/JSONEncoder.java [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+        //string.deleteCharAt(string.length() - 3);
+        string.append(getTabs(tabs + 1));
+        string.append("]\n");
+        return string;
+    }
+
+    private StringBuilder getTabs(int tabs) {
+        StringBuilder tab = new StringBuilder();
+        while (tabs != 0) {
+            tab.append("\t");
```
Comment (roman-mate) [inline serialization/encoder/JSONEncoder.java]: spaces instead of \t

File: serialization/encoder/JSONEncoder.java [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+        switch (fieldName) {
+            case "name":
+                string.append(getNameFieldLine(group, tabs));
+                return string;
+            case "figures":
+                string.append(getFigureListItemsLine(group, tabs));
+                return string;
+            case "groups":
+                string.append(getGroupListItemsLine(group, tabs));
+                return string;
```
Comment (roman-mate) [inline serialization/encoder/JSONEncoder.java]: always have default.
https://mate-academy.github.io/style-guides/java/java.html#s4.8.4-switch

"Each switch statement includes a default statement group, even if it contains no code."

File: serialization/encoder/XMLEncoder.java [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+        return string;
+    }
+
+    private StringBuilder getGroupXMLCode(Group group, Class clazz, int tabs) {
+        StringBuilder string = new StringBuilder();
+
+        string.append(getObjectClassLine(clazz, tabs));
+
+        string.append(getObjectFieldsLine(group, clazz, tabs + 1));
+        string.append(getObjectMethodLine(clazz, tabs + 1));
```
Comment (roman-mate) [inline serialization/encoder/XMLEncoder.java]: no need to serialize methods. We serialize only data stored in the object..

File: serialization/encoder/XMLEncoder.java [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+
+        string.append(getTabs(tabs));
+        string.append("</object>\n");
+        return string;
+    }
+
+    private StringBuilder getObjectMethodLine(Class clazz, int tabs) {
```
Comment (roman-mate) [inline serialization/encoder/XMLEncoder.java]: suggestion: getObjectMethodLine => generateObjectMethodLine

Same everywhere.

File: files/second_group.json [hunk 28]
```diff
@@ -0,0 +1,38 @@
+{
+	"name":"Group_2",
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply files/second_group.json]: refactored

File: files/second_group.xml [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+<java version="10.0.2" class="serialization.Group">
+	<object class="Group" id="42121758">
+		<void property="name">
+			<string>VALUE</string>
+		</void>
+		<void property="figures">
+			<list>VALUE</list>
+				<object class="Triangle" id="257895351">
+					<void property="name">
+						<string>VALUE</string>
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply files/second_group.xml]: refactored

File: files/second_group.xml [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+					<void property="b">
+						<int>VALUE</int>
+					</void>
+					<void property="c">
+						<int>VALUE</int>
+					</void>
+					<void property="square">
+						<double>VALUE</double>
+					</void>
+					<void method="getName"> </void>
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply files/second_group.xml]: refactored

File: immutableclass/ImmutableAnimal.java [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+public final class ImmutableAnimal {
+
+    private final String name;
+    private final int age;
+    private final Family family;
+
+    public ImmutableAnimal(String name, int age, Family family) {
+        this.name = name;
+        this.age = age;
+        //this.family = family;
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply immutableclass/ImmutableAnimal.java]: refactored

File: immutableclass/ImmutableAnimal.java [hunk 32]
```diff
@@ -0,0 +1,56 @@
...
+        Family familyClone = new Family(family.getFather(), family.getMother());
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply immutableclass/ImmutableAnimal.java]: refactored

File: immutableclass/ImmutableAnimal.java [hunk 33]
```diff
@@ -0,0 +1,56 @@
...
+        return age;
+    }
+
+    public Family getFamily() {
+        //return family;
+        return new Family(this.family.getFather(), this.family.getMother());
+    }
+}
+
+class Family {
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply immutableclass/ImmutableAnimal.java]: refactored

File: immutableclass/ImmutableAnimal.java [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    String getMother() {
+        return mother;
+    }
+
+    void setMother(String mother) {
+        this.mother = mother;
+    }
+}
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply immutableclass/ImmutableAnimal.java]: refactored

File: immutableclass/ImmutableClassTest.java (hunk 13 above)
Comment (Osievskaya23) [reply immutableclass/ImmutableClassTest.java]: refactored

File: patterns/composite/Team.java [hunk 35]
```diff
@@ -0,0 +1,23 @@
+package patterns.composite;
+
+import java.util.ArrayList;
+import java.util.List;
+
+class Team {
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply patterns/composite/Team.java]: refactored

File: patterns/composite/Team.java [hunk 36]
```diff
@@ -0,0 +1,23 @@
...
+    private List<Developer> developers = new ArrayList<>();
+
+    void addDeveloper(Developer developer) {
+        developers.add(developer);
+    }
+
+    void removeDeveloper(Developer developer) {
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply patterns/composite/Team.java]: refactored

File: serialization/Circle.java (hunk 14 above)
Comment (Osievskaya23) [reply serialization/Circle.java]: refactored

File: serialization/Figure.java (hunk 15 above)
Comment (Osievskaya23) [reply serialization/Figure.java]: refactored

File: serialization/Geometry.java [hunk 37]
```diff
@@ -0,0 +1,41 @@
...
+        Figure thirdSquare = new Square("Square_3", 6);
+
+        firstGroup.addFigures(firstTriangle, firstCircle, firstSquare);
+
+        secondGroup.addFigures(secondTriangle, secondSquare, thirdSquare);
+        secondGroup.addGroups(firstGroup);
+
+        secondGroup.getFigureSquares();
+
+        XMLEncoder xmlEncoder = new XMLEncoder();
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply serialization/Geometry.java]: Refactored

File: serialization/Group.java (hunk 16 above)
Comment (Osievskaya23) [reply serialization/Group.java]: refactored

File: serialization/Group.java [hunk 38]
```diff
@@ -0,0 +1,50 @@
...
+
+    public void addFigures(Figure ... neadAddingFigures) {
+        figures.addAll(Arrays.asList(neadAddingFigures));
+    }
+
+    public void addGroups(Group ... neadAddGroups) {
+        groups.addAll(Arrays.asList(neadAddGroups));
+    }
+
+    public void getFigureSquares() {
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply serialization/Group.java]: refactored

File: serialization/Square.java [hunk 39]
```diff
@@ -0,0 +1,26 @@
+package serialization;
+
+public class Square implements Figure {
+    public String name;
+    private int a;
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply serialization/Square.java]: refactored

File: serialization/Triangle.java [hunk 40]
```diff
@@ -0,0 +1,39 @@
+package serialization;
+
+public class Triangle implements Figure {
+    public String name;
+    private int a;
+    private int b;
+    private int c;
+    public double square;
+
+    public Triangle(String name, int a, int b, int c) {
```
[the remark this answers is not in the document]
Comment (Osievskaya23) [reply serialization/Triangle.java]: refactored

File: serialization/encoder/JSONEncoder.java (hunk 25 above)
Comment (Osievskaya23) [reply serialization/encoder/JSONEncoder.java]: refactored

File: com/serialization/Circle.java [hunk 41]
```diff
@@ -0,0 +1,27 @@
+package com.serialization;
+
+public class Circle implements Figure {
+    public static final double PI = 3.141_592_653_589_793;
```
Comment (roman-mate) [inline com/serialization/Circle.java]: Just realized there is Math.PI. So you can get rid of this constant.

File: com/serialization/Circle.java [hunk 42]
```diff
@@ -0,0 +1,27 @@
...
+    private String name;
+    private int radius;
+    public double square;
```
Comment (roman-mate) [inline com/serialization/Circle.java]: Why public? You have public getter to access square.

File: com/serialization/Geometry.java [hunk 43]
```diff
@@ -0,0 +1,42 @@
+package com.serialization;
+
+import com.serialization.encoder.JsonEncoder;
+import com.serialization.encoder.XmlEncoder;
+
+import java.io.File;
+
+public class Geometry {
+    private final static String PATH_XML = "C:\\Users\\38066\\IdeaProjects\\GoodPractice\\src\\files\\group.xml";
```
Comment (roman-mate) [inline com/serialization/Geometry.java]: suggestion: DEFAULT_PATH_XML. You might want to get rid of absolute path and use relative path instead. src/files/group.xml

File: com/serialization/Square.java [hunk 44]
```diff
@@ -0,0 +1,26 @@
+package com.serialization;
+
+public class Square implements Figure {
+    public String name;
```
Comment (roman-mate) [inline com/serialization/Square.java]: all fields pretty much always should be private and you should have getters for them.

File: com/serialization/Triangle.java [hunk 45]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Override
+    public double getSquare() {
+        return square;
+    }
+
+    @Override
+    public double calculateSquare() {
+        double square = 0;
+        if(firstSideLength + secondSideLength > thirdSideLength &&
```
Comment (roman-mate) [inline com/serialization/Triangle.java]: typically we put && on the new line.

File: com/serialization/Triangle.java [hunk 46]
```diff
@@ -0,0 +1,41 @@
...
+                firstSideLength + thirdSideLength > secondSideLength &&
+                secondSideLength + thirdSideLength > firstSideLength) {
+            int p = (firstSideLength + secondSideLength + thirdSideLength) / 2;
```
Comment (roman-mate) [inline com/serialization/Triangle.java]: p => perimeter

**Pull request merged**